chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+}
+```
+
+To provide the necessary references, NLS makes used of the *scope safe* memory of its `Linearizer` implementation.
+This is possible, because each record value corresponds to its own scope.
+The complete process looks as follows:
+
+1. When registering a record, first the outer `Record` is added to the linearization
+2. This is followed by `RecordField` items for its fields, which at this point do not reference any value.
+3. NLS then stores the `id` of the parent as well as the fields and the offsets of the corresponding items (`n-4` and `[(apiVersion, n-3), (containers, n-2), (metadata, n-1)]` respectively in the example [@fig:nls-lin-records]).
+4. The `scope` method will be called in the same order as the record fields appear.
+   Using this fact, the `scope` method moves the data stored for the next evaluated field into the freshly generated `Linearizer`
+5. **(In the sub-scope)** The `Linearizer` associates the `RecordField` item with the (now known) `id` of the field's value.
+   The cached field data is invalidated such that this process only happens once for each field.
+
+
+##### Variable Reference
+
+The usage of a variable is always expressed as a `Var` node that holds an identifier.
+Registering a name usage is a multi-step process.
+
+First, NLS tries to find the identifier in its scope-aware name registry.
+If the registry does not contain the identifier, NLS will linearize the node as `Unbound`.
+In the case that the registry lookup succeeds, NLS retrieves the referenced `Declaration` or `RecordField`. The linearizer will then add a usage item in the `Resolved` state to the linearization and update the declaration's list of usages.
+
+##### Resolution of Record Fields
+
+The AST representation of record destructuring in [@fig:nickel-static-access] shows that accessing inner records involves chains of unary operations *ending* with a reference to a variable binding.
+Each operation encodes one field of a referenced record.
+However, to reference the corresponding declaration, the final usage has to be known.
+Therefore, instead of linearizing the intermediate elements directly, the `Linearizer` adds them to a shared stack until the grounding variable reference is registered.
+
+Whenever a variable usage is linearized, NLS checks the stack for latent destructors.
+If destructors are present, it adds `Usage` items for each element on the stack.
+Yet, because records are recursive it is possible that fields reference other fields' values.
+
+Consider the following example [@lst:nickel-recursive-record], which is depicted in [@fig:nls-unavailable-rec-record-field]
+
+
+```{.nickel #lst:nickel-recursive-record caption="Example of a recursive record"}
+{
+  y = {
+    yy = "foo",
+    yz = z,
+  },
+  z = y.yy
+}
+```
+
+```{.graphviz #fig:nls-unavailable-rec-record-field caption="Example race condition in recursive records. The field `y.yz` cannot be not be referenced at this point as the `y` branch has yet to be linearized"}
```

Comment:
```suggestion
```{.graphviz #fig:nls-unavailable-rec-record-field caption="Example lock in recursive records. The field `y.yz` cannot be not be referenced at this point as the `y` branch has yet to be linearized"}
```

I would say a race condition is something a bit different.